cmd/kubeadm/app/cmd/join.go

```diff
@@ -647,7 +647,7 @@ func fetchInitConfiguration(tlsBootstrapCfg *clientcmdapi.Config) (*kubeadmapi.I
 	}
 
 	// Fetches the init configuration
-	initConfiguration, err := configutil.FetchConfigFromFileOrCluster(tlsClient, os.Stdout, "join", "", true)
+	initConfiguration, err := configutil.FetchConfigFromCluster(tlsClient, os.Stdout, "join", true)
```

Comment:
Ideally `FetchConfigFromCluster` would be renamed to `FetchClusterConfiguration` and would return ClusterConfiguration only. `getNodeRegistration` and `getAPIEndpoint` will become public separate functions.
Another way to tackle this is to introduce an interface, that has `GetClientConfiguration()`, `GetNodeRegistration()` and `GetLocalAPIEndpoint()`. InitConfiguration will implement this, so it will be straightforward to use. Also, there will be an internal implementation, that takes kubeconfig and/or JoinConfiguration and then implement the interface by fetching stuff from the cluster or JoinConfiguration. This approach has the benefit of replacing InitConfiguration in most places with a single type and of adding caching (so that we fetch ClusterConfiguration only once for example).
I am experimenting with both approaches, but the first code will probably become available in the 1.15 time frame.

Reply:
> became resilient to the fact the kube-proxy add-on exists or not (if I'm not wrong @neolit123 is working on this)

waiting on the join phases work to finish.

but i think we might have a problem related to that.
i think the original reason to fetch the component config for kubeproxy was to be able to perform specific preflight checks on the joining node (was it about IPVS?). if we fetch cluster config only, these preflight checks will have to be removed.